File: tests/example-frames/deformers/skin-vertex.rs
// $dep skinned_human.fbx
// $crate glam
// $ cargo build
// $ cargo run --quiet -- skinned_human.fbx Female > result.obj
use ufbx;
use glam::*;

trait AsGlam<T> {
    fn as_glam(self) -> T;
}

impl AsGlam<DVec3> for ufbx::Vec3 {
    fn as_glam(self) -> DVec3 {
        DVec3::new(self.x, self.y, self.z)
    }
}

impl AsGlam<DMat4> for ufbx::Matrix {
    fn as_glam(self) -> DMat4 {
        DMat4::from_cols(
            DVec4::new(self.m00, self.m10, self.m20, 0.0),
            DVec4::new(self.m01, self.m11, self.m21, 0.0),
            DVec4::new(self.m02, self.m12, self.m22, 0.0),
            DVec4::new(self.m03, self.m13, self.m23, 1.0))
    }
}

// -- EXAMPLE_SOURCE --

fn main() {
    let path = std::env::args().nth(1).expect("expected .fbx path");
    let node_name = std::env::args().nth(2).expect("expected node name");

    let scene = ufbx::load_file(&path, Default::default()).expect("failed to load scene");
    let node = scene.find_node(&node_name).expect("could not find node");
    let mesh = node.mesh.as_ref().expect("node did not contain a mesh");
    let skin = mesh.skin_deformers.get(0).expect("no skin deformer in mesh");
    for i in 0..mesh.num_vertices {
        let vertex = transform_vertex(&mesh, &skin, i);
        println!("v {:.5} {:.5} {:.5}", vertex.x, vertex.y, vertex.z);
    }

    // Output faces
    for face in &mesh.faces {
        print!("f");
        for j in 0..face.num_indices {
            print!(" {}", mesh.vertex_indices[(face.index_begin + j) as usize] + 1);
        }
        println!("");
    }
}

File: tests/example-frames/deformers/skin-vertex.cpp
// $dep ufbx example_base skinned_human.fbx
// $ clang++ ufbx.c main.cpp -lm -o example
// $ ./example skinned_human.fbx Female > result.obj
#include "ufbx.h"
#include "example_base.h"
#include <stdio.h>

// -- EXAMPLE_SOURCE --

int main(int argc, char **argv)
{
    ufbx_scene *scene = ufbx_load_file(argv[1], NULL, NULL);
    assert(scene);

    ufbx_node *node = ufbx_find_node(scene, argv[2]);
    assert(node && node->mesh);

    ufbx_mesh *mesh = node->mesh;
    assert(mesh->skin_deformers.count > 0);
    ufbx_skin_deformer *skin = mesh->skin_deformers[0];

    for (uint32_t i = 0; i < mesh->num_vertices; i++) {
        Vector3 vertex = transform_vertex(mesh, skin, i);
        printf("v %.5f %.5f %.5f\n", vertex.x, vertex.y, vertex.z);
    }

    // Output faces
    for (ufbx_face face : mesh->faces) {
        printf("f");
        for (uint32_t j = 0; j < face.num_indices; j++) {
            printf(" %u", mesh->vertex_indices[face.index_begin + j] + 1);
        }
        printf("\n");
    }

    return 0;
}
